goproxy.go

```diff
@@ -193,21 +193,31 @@ func (g *Goproxy) load() {
 		g.goBinEnv[parts[0]] = parts[1]
 	}
 
-	var proxies []string
-	for _, proxy := range strings.Split(g.goBinEnv["GOPROXY"], ",") {
-		proxy = strings.TrimSpace(proxy)
-		if proxy == "" {
-			continue
-		}
+	var proxyGroup [][]string
+	for _, proxies := range strings.Split(g.goBinEnv["GOPROXY"], "|") {
+		var cleanProxies []string
+		for _, proxy := range strings.Split(proxies, ",") {
+			proxy = strings.TrimSpace(proxy)
+			if proxy == "" {
+				continue
+			}
 
-		proxies = append(proxies, proxy)
-		if proxy == "direct" || proxy == "off" {
-			break
+			cleanProxies = append(cleanProxies, proxy)
+			if proxy == "direct" || proxy == "off" {
+				break
+			}
+		}
+		if len(cleanProxies) > 0 {
+			proxyGroup = append(proxyGroup, cleanProxies)
 		}
 	}
 
-	if len(proxies) > 0 {
-		g.goBinEnv["GOPROXY"] = strings.Join(proxies, ",")
+	if len(proxyGroup) > 0 {
+		var cleanProxyGroup []string
+		for _, proxies := range proxyGroup {
+			cleanProxyGroup = append(cleanProxyGroup, strings.Join(proxies, ","))
+		}
+		g.goBinEnv["GOPROXY"] = strings.Join(cleanProxyGroup, "|")
```

Comment:
According to `go help modules`:

> ...
> If GOPROXY is set to the string "direct", downloads use a direct connection to
> source control servers. Setting GOPROXY to "off" disallows downloading modules
> from any source. Otherwise, GOPROXY is expected to be list of module proxy URLs
> separated by either comma (,) or pipe (|) characters, which control error
> fallback behavior. For each request, the go command tries each proxy in
> sequence. If there is an error, the go command will try the next proxy in the
> list if the error is a 404 or 410 HTTP response or if the current proxy is
> followed by a pipe character, indicating it is safe to fall back on any error.
> ...

we should regard comma (,) and pipe (|) as separators of the **same level**, which means that there is no "containment" relationship between them.

And the code block here is mainly used to format the `GOPROXY` environment variable, such as remove spaces, truncate unreachable fields, etc. So I think maybe we can find a better way than using `strings.Split` here, like this:

```go
var envGOPROXY string
for goproxy := g.goBinEnv["GOPROXY"]; goproxy != ""; {
	var proxy, sep string
	if i := strings.IndexAny(goproxy, ",|"); i >= 0 {
		proxy = goproxy[:i]
		sep = string(goproxy[i])
		goproxy = goproxy[i+1:]
		if goproxy == "" {
			sep = ""
		}
	} else {
		proxy = goproxy
		goproxy = ""
	}

	proxy = strings.TrimSpace(proxy)
	switch proxy {
	case "":
		continue
	case "direct", "off":
		sep = ""
		goproxy = ""
	}

	envGOPROXY = fmt.Sprint(envGOPROXY, proxy, sep)
}

if envGOPROXY != "" {
	g.goBinEnv["GOPROXY"] = envGOPROXY
} else if g.goBinEnv["GOPROXY"] == "" {
	g.goBinEnv["GOPROXY"] = "https://proxy.golang.org,direct"
} else {
	g.goBinEnv["GOPROXY"] = "off"
}
```

Reply:
I just fixed this with the 871b5310e549c305487f4d98a4ae10b8a7c8ba00.